### PR TITLE
many: use laid out types only for writing content

### PR DESCRIPTION
**Pull request by alfonsosanchezbeato**
Use laid out types only for writing content on installations and factory reset.

### CODE REVIEW

File: gadget/layout.go [hunk 1]
```diff
@@ -354,6 +334,49 @@ func LayoutVolume(volume *Volume, gadgetToDiskStruct map[int]*OnDiskStructure, o
...
+			return err
+		}
+	}
+
+	los.LaidOutContent = content
+	los.ResolvedContent = resolvedContent
+	return nil
+}
+
+func LayoutVolumeStructure(onDisk *OnDiskStructure, vs *VolumeStructure, kernelInfo *kernel.Info, opts *LayoutOptions) (*LaidOutStructure, error) {
```
[the remark this answers is not in the document]
Comment (pedronis) [reply gadget/layout.go:367]: also this does not seem to have unit tests of its own

File: gadget/gadget.go [hunk 2]
```diff
@@ -185,7 +185,15 @@ func (v *Volume) MinSize() quantity.Size {
 	return quantity.Size(endVol)
 }
 
-func (v *Volume) yamlIdxToStructureIdx(yamlIdx int) (int, error) {
+func (v *Volume) StructFromYamlIndex(yamlIdx int) *VolumeStructure {
```
Comment (pedronis) [inline gadget/gadget.go:188]: this doesn't seem to be reached by tests here

File: gadget/gadget.go [hunk 3]
```diff
@@ -1675,6 +1684,18 @@ func LaidOutVolumesFromGadget(gadgetRoot, kernelRoot string, model Model, encTyp
 	return system, all, nil
 }
 
+// FindBootVolume returns the volume that contains the system-boot partition.
+func FindBootVolume(vols map[string]*Volume) (*Volume, error) {
```
Comment (pedronis) [inline gadget/gadget.go:1688]: this is not tested

File: gadget/install/partition.go [hunk 4]
```diff
@@ -167,58 +165,78 @@ func buildPartitionList(dl *gadget.OnDiskVolume, lov *gadget.LaidOutVolume, opts
...
 		node := deviceName(dl.Device, pIndex)
 		// Change bits that depend on the disk, which includes
 		// overriding the size from the gadget as we might be
 		// expanding the data partition.
-		// TODO fill in construction of LaidOutStructure instead if/when possible
-		laidOut.Node = node
-		laidOut.DiskIndex = pIndex
-		laidOut.Size = quantity.Size(newSizeInSectors * sectorSize)
+		vs.Device = node
+		vs.Size = quantity.Size(newSizeInSectors * sectorSize)
```
Comment (pedronis) [inline gadget/install/partition.go:224]: again it's a bit murky to modify the input in place now, and it also gets reflected into the toBeCreated entries

File: gadget/install/install.go [hunk 5]
```diff
@@ -96,25 +100,25 @@ func saveStorageTraits(mod gadget.Model, allLaidOutVols map[string]*gadget.LaidO
 	return nil
 }
 
-func maybeEncryptPartition(laidOut *gadget.LaidOutStructure, encryptionType secboot.EncryptionType, sectorSize quantity.Size, perfTimings timings.Measurer) (fsParams *mkfsParams, encryptionKey keys.EncryptionKey, err error) {
+func maybeEncryptPartition(diskPart *gadget.OnDiskStructure, volStruct *gadget.VolumeStructure, encryptionType secboot.EncryptionType, sectorSize quantity.Size, perfTimings timings.Measurer) (fsParams *mkfsParams, encryptionKey keys.EncryptionKey, err error) {
```
Comment (pedronis) [inline gadget/install/install.go:103]: why this change, shouldn't diskPart reflect volStruct anyway?

File: gadget/install/install.go [hunk 6]
```diff
@@ -182,22 +187,31 @@ func writePartitionContent(laidOut *gadget.LaidOutStructure, fsDevice string, ob
 	return nil
 }
 
-func installOnePartition(laidOut *gadget.LaidOutStructure, encryptionType secboot.EncryptionType, sectorSize quantity.Size, observer gadget.ContentObserver, perfTimings timings.Measurer) (fsDevice string, encryptionKey keys.EncryptionKey, err error) {
+func installOnePartition(diskPart *gadget.OnDiskStructure, vs *gadget.VolumeStructure, kernelInfo *kernel.Info, gadgetRoot, kernelRoot string, encryptionType secboot.EncryptionType, sectorSize quantity.Size, observer gadget.ContentObserver, perfTimings timings.Measurer) (fsDevice string, encryptionKey keys.EncryptionKey, err error) {
```
Comment (pedronis) [inline gadget/install/install.go:190]: same question

File: gadget/install/install.go [hunk 7]
```diff
@@ -618,28 +682,22 @@ func SaveStorageTraits(model gadget.Model, allLaidOutVols map[string]*gadget.Lai
...
-
 	setupData := &EncryptionSetupData{
 		parts: make(map[string]partEncryptionData),
 	}
 	for volName, vol := range onVolumes {
-		var onDiskVol *gadget.OnDiskVolume
+		onDiskVol, err := OnDiskVolumeFromGadgetVol(vol)
+		if err != nil {
+			return nil, err
+		}
```
Comment (pedronis) [inline gadget/install/install.go:700]: shouldn't we do this only as required below?

File: gadget/install/install.go [hunk 8]
```diff
@@ -492,7 +566,12 @@ func WriteContent(onVolumes map[string]*gadget.Volume, allLaidOutVols map[string
 
 	var onDiskVols []*gadget.OnDiskVolume
 	for volName, vol := range onVolumes {
-		var onDiskVol *gadget.OnDiskVolume
+		onDiskVol, err := OnDiskVolumeFromGadgetVol(vol)
```
Comment (pedronis) [inline gadget/install/install.go:569]: same

File: gadget/install/install.go [hunk 9]
```diff
@@ -307,11 +326,20 @@ func createEncryptionParams(encTyp secboot.EncryptionType) gadget.StructureEncry
 	return gadget.StructureEncryptionParameters{}
 }
 
+func onDiskStructsSortedIdx(vss map[int]*gadget.OnDiskStructure) []int {
```
Comment (pedronis) [inline gadget/install/install.go:329]: this could return perhaps the sorted structures instead of just the indexes, or use a channel to do that as an iterator

**Review by pedronis — COMMENTED** [review]
did a first pass, the comments are a bit of various kinds, I honestly struggled a bit reading through the changes to install, I also remarked in a couple of points that the functions now updating gadget.Volume in place are maybe not the cleanest and maybe they should return modified copies instead

File: gadget/install/install.go (hunk 5 above)
Comment (pedronis) [inline gadget/install/install.go:103]: do we need a new structure in the code here to pair volStruct nad diskPart?

File: overlord/devicestate/handlers_install.go [hunk 10]
```diff
@@ -949,42 +950,75 @@ func (m *DeviceManager) doInstallFinish(t *state.Task, _ *tomb.Tomb) error {
...
+				"ubuntu-data": {Method: gadget.EncryptionLUKS},
+				"ubuntu-save": {Method: gadget.EncryptionLUKS},
+			}
+		}
+		gadgetToDiskMap, err := gadget.EnsureVolumeCompatibility(vol, diskVolume, volCompatOpts)
+		if err != nil {
+			return err
+		}
+		volToGadgetToDiskStruct[name] = gadgetToDiskMap
+	}
```
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply overlord/devicestate/handlers_install.go]: I've moved this now

**Review by alfonsosanchezbeato — COMMENTED** [review]
Thanks for the review, this is ready now for another pass

File: daemon/api_debug.go [hunk 11]
```diff
@@ -284,53 +281,24 @@ func getGadgetDiskMapping(st *state.State) Response {
...
-		default:
-			return InternalError("unknown sealing method: %s", sealingMethod)
-		}
-	}
-
-	_, allLaidOutVols, err := gadget.LaidOutVolumesFromGadget(gadgetDir, kernelDir, mod, encType)
+	info, err := gadget.ReadInfoAndValidate(gadgetDir, mod, nil)
 	if err != nil {
-		return InternalError("cannot get all disk volume device traits: cannot layout volumes: %v", err)
+		return InternalError("cannot get all disk volume device traits: while reading gadget: %v", err)
```
Comment (pedronis) [inline daemon/api_debug.go:288]: shouldn't this be "...: cannot read gadget: ..." ?

File: gadget/gadget.go [hunk 12]
```diff
@@ -195,6 +207,41 @@ func (v *Volume) yamlIdxToStructureIdx(yamlIdx int) (int, error) {
...
+				offsetWr := *vs.OffsetWrite
+				newVs.OffsetWrite = &offsetWr
+			}
+			if vs.Content != nil {
+				newVs.Content = make([]VolumeContent, len(vs.Content))
+				copy(newVs.Content, vs.Content)
+				for i, c := range vs.Content {
+					if c.Offset != nil {
+						newC := &newVs.Content[i]
+						newC.Offset = asOffsetPtr(*c.Offset)
```
Comment (pedronis) [inline gadget/gadget.go:235]: this code does not seem reached by tests

File: gadget/layout.go [hunk 13]
```diff
@@ -276,16 +279,14 @@ func layoutVolumePartially(volume *Volume, gadgetToDiskStruct map[int]*OnDiskStr
...
-	for i := range los {
-		los[i].PartitionFSLabel = los[i].Label()
-		los[i].PartitionFSType = los[i].Filesystem()
-		if encType != secboot.EncryptionTypeNone {
-			switch los[i].Role() {
-			case SystemData, SystemSave:
-				los[i].PartitionFSLabel += "-enc"
-				los[i].PartitionFSType = "crypto_LUKS"
-			}
+func setOnDiskLabelAndTypeInLaidOuts(los *LaidOutStructure, encType secboot.EncryptionType) {
```
Comment (pedronis) [inline gadget/layout.go:282]: s/LaidOuts/LaidOut/ ?

File: gadget/ondisk.go [hunk 14]
```diff
@@ -188,3 +195,70 @@ func OnDiskStructureFromPartition(p disks.Partition) (OnDiskStructure, error) {
 		Node:             p.KernelDeviceNode,
 	}, nil
 }
+
+// OnDiskVolumeFromGadgetVol returns the disk volume matching a gadget volume
+// that has the Device field set, which implies that this should be called only
+// in the context of an installer that set the device in the gadget and
+// returned it to snapd.
+func OnDiskVolumeFromGadgetVol(vol *Volume) (*OnDiskVolume, error) {
```
Comment (pedronis) [inline gadget/ondisk.go:203]: this needs its own tests here

File: gadget/install/install_dummy.go [hunk 15]
```diff
@@ -63,3 +59,7 @@ func EncryptPartitions(onVolumes map[string]*gadget.Volume, encryptionType secbo
 func KeysForRole(setupData *EncryptionSetupData) map[string]keys.EncryptionKey {
 	return nil
 }
+
+func OnDiskVolumeFromGadgetVol(vol *gadget.Volume) (*gadget.OnDiskVolume, error) {
+	return nil, fmt.Errorf("build without secboot support")
+}
```
Comment (pedronis) [inline gadget/install/install_dummy.go:65]: should this go away now?

File: gadget/install/install.go [hunk 16]
```diff
@@ -809,3 +819,23 @@ func FactoryReset(model gadget.Model, gadgetRoot, kernelRoot, bootDevice string,
 		DeviceForRole: deviceForRole,
 	}, nil
 }
+
+// MatchDisksToGadgetVolumes matches gadget volumes with disks present in the
+// system, taking into account the provided compatibility options.
```
Comment (pedronis) [inline gadget/install/install.go:824]: this comment should described explicitly what are the keys of the returned map.

File: gadget/install/install.go [hunk 17]
```diff
@@ -229,70 +247,75 @@ func resolveBootDevice(bootDevice string, bootVol *gadget.Volume) (string, error
...
 	if model.Grade() == asserts.ModelGradeUnset {
-		return "", nil, nil, 0,
+		return "", nil, 0,
 			fmt.Errorf("cannot run install mode on pre-UC20 system")
 	}
 
-	// Find boot laid out data
-	laidOutBootVol, allLaidOutVols, err := gadget.LaidOutVolumesFromGadget(gadgetRoot, kernelRoot, model, options.EncryptionType)
+	// Find boot volume
+	info, err := gadget.ReadInfoAndValidate(gadgetRoot, model, nil)
```
Comment (pedronis) [inline gadget/install/install.go:267]: should the function receive this from the caller, Run needs to do this too atm

File: gadget/install/install.go [hunk 18]
```diff
@@ -96,25 +100,27 @@ func saveStorageTraits(mod gadget.Model, allLaidOutVols map[string]*gadget.LaidO
 	return nil
 }
 
-func maybeEncryptPartition(laidOut *gadget.LaidOutStructure, encryptionType secboot.EncryptionType, sectorSize quantity.Size, perfTimings timings.Measurer) (fsParams *mkfsParams, encryptionKey keys.EncryptionKey, err error) {
+func maybeEncryptPartition(pair *gadget.OnDiskAndGadgetStructurePair, encryptionType secboot.EncryptionType, sectorSize quantity.Size, perfTimings timings.Measurer) (fsParams *mkfsParams, encryptionKey keys.EncryptionKey, err error) {
```
Comment (pedronis) [inline gadget/install/install.go:103]: should we use d(isk)g(adget)pair dgpair instead of just pair here and other places?

**Review by pedronis — COMMENTED** [review]
did another pass

**Review by alfonsosanchezbeato — COMMENTED** [review]
Thanks for the review, I've pushed changes now

File: gadget/gadget.go [hunk 19]
```diff
@@ -185,7 +185,19 @@ func (v *Volume) MinSize() quantity.Size {
...
+	i, err := v.YamlIdxToStructureIdx(yamlIdx)
+	if err != nil {
+		return nil
+	}
+	return &v.Structure[i]
+}
+
+// YamlIdxToStructureIdx returns the index to Volume.Structure that matches the
+// yaml index from the original yaml file.
+func (v *Volume) YamlIdxToStructureIdx(yamlIdx int) (int, error) {
```
Comment (pedronis) [inline gadget/gadget.go:200]: afaics this is not used outside of gadget so maybe it could stay unexported?

File: overlord/devicestate/handlers_install.go [hunk 20]
```diff
@@ -949,42 +950,68 @@ func (m *DeviceManager) doInstallFinish(t *state.Task, _ *tomb.Tomb) error {
 		return err
 	}
 
+	gi, err := gadget.ReadInfoAndValidate(mntPtForType[snap.TypeGadget], sys.Model, nil)
+	if err != nil {
+		return err
+	}
+
```
Comment (pedronis) [inline overlord/devicestate/handlers_install.go:957]: should the TODO:

// TODO validation of onVolumes versus gadget.yaml, considering also partial

get closer to the code here now?

or is it actually done because MatchDisksToGadgetVolumes below now uses EnsureVolumeCompatibility ?

File: gadget/install/install.go [hunk 21]
```diff
@@ -307,35 +312,64 @@ func createEncryptionParams(encTyp secboot.EncryptionType) gadget.StructureEncry
...
+		return nil, err
+	}
 	hasSavePartition := false
-
 	// Note that all partitions here will have a role (see
-	// gadget.IsCreatableAtInstall() which defines the list)
-	for _, laidOut := range created {
+	// gadget.IsCreatableAtInstall() which defines the list). We do it in
+	// the order in which partitions were specified in the gadget.
+	for _, pair := range created {
```
Comment (pedronis) [inline gadget/install/install.go:362]: s/pair/dgpair/ ?

File: gadget/layout.go
Comment (pedronis) [inline gadget/layout.go:1]: there are some XXXs in this file, especially the TEMPORARY one that I'm not sure if/when they will be addressed

File: gadget/install/install.go [hunk 22]
```diff
@@ -809,3 +809,25 @@ func FactoryReset(model gadget.Model, gadgetRoot, kernelRoot, bootDevice string,
...
+	volCompatOpts *gadget.VolumeCompatibilityOptions) (map[string]map[int]*gadget.OnDiskStructure, error) {
+	volToGadgetToDiskStruct := map[string]map[int]*gadget.OnDiskStructure{}
+	for name, vol := range gVols {
+		diskVolume, err := gadget.OnDiskVolumeFromGadgetVol(vol)
+		if err != nil {
+			return nil, err
+		}
+		gadgetToDiskMap, err := gadget.EnsureVolumeCompatibility(vol, diskVolume, volCompatOpts)
+		if err != nil {
+			return nil, err
```
Comment (pedronis) [inline gadget/install/install.go:827]: should this error path be exercised by tests?

**Review by pedronis — COMMENTED** [review]
thanks for the changes, overall looks good, did another pass

File: gadget/layout.go
Comment (alfonsosanchezbeato) [reply gadget/layout.go:1]: I've expanded a bit the comments on this, as discussed.

File: overlord/devicestate/handlers_install.go (hunk 20 above)
Comment (alfonsosanchezbeato) [reply overlord/devicestate/handlers_install.go:955]: I've moved the TODO down, a compatibility check between the gadget and what we take from the installer needs still to happen.

**Review by alfonsosanchezbeato — COMMENTED** [review]
Thanks, I've pushed now changes

**Review by pedronis — APPROVED** [review]
thank you

File: gadget/install/partition.go [hunk 23]
```diff
@@ -167,58 +165,78 @@ func buildPartitionList(dl *gadget.OnDiskVolume, lov *gadget.LaidOutVolume, opts
 
 	// Check if the last partition has a system-data role
 	canExpandData := false
-	if n := len(lov.LaidOutStructure); n > 0 {
-		last := lov.LaidOutStructure[n-1]
-		if last.Role() == gadget.SystemData {
+	if n := len(vol.Structure); n > 0 {
+		last := vol.Structure[n-1]
+		if last.Role == gadget.SystemData {
 			canExpandData = true
```
Comment (Meulengracht) [inline gadget/install/partition.go:171]: missing `break`?

File: gadget/layout.go [hunk 24]
```diff
@@ -321,29 +323,8 @@ func LayoutVolume(volume *Volume, gadgetToDiskStruct map[int]*OnDiskStructure, o
...
-		}
-
-		// resolve filesystem content
-		if doResolveContent {
-			resolvedContent, err := resolveVolumeContent(opts.GadgetRootDir, opts.KernelRootDir, kernelInfo, &structures[idx], nil)
-			if err != nil {
-				return nil, err
-			}
-			structures[idx].ResolvedContent = resolvedContent
+		if err := fillLaidoutStructure(&structures[idx], kernelInfo, opts); err != nil {
```
Comment (Meulengracht) [inline gadget/layout.go:326]: this is a nice change!

File: gadget/ondisk.go [hunk 25]
```diff
@@ -188,3 +195,70 @@ func OnDiskStructureFromPartition(p disks.Partition) (OnDiskStructure, error) {
...
+	onDiskVol, err := OnDiskVolumeFromDisk(disk)
+	if err != nil {
+		return nil, fmt.Errorf("cannot retrieve on disk volume for %q: %v", partPath, err)
+	}
+
+	return onDiskVol, nil
+}
+
+func MockSysfsPathForBlockDevice(f func(device string) (string, error)) (restore func()) {
+	old := sysfsPathForBlockDevice
```
Comment (Meulengracht) [inline gadget/ondisk.go:259]: nitpick: `testutil.backup`?

File: overlord/devicestate/export_test.go [hunk 26]
```diff
@@ -403,14 +403,23 @@ func MockInstallEncryptPartitions(f func(onVolumes map[string]*gadget.Volume, en
...
+func MockInstallSaveStorageTraits(f func(model gadget.Model, allLaidOutVols map[string]*gadget.Volume, encryptSetupData *install.EncryptionSetupData) error) (restore func()) {
 	old := installSaveStorageTraits
 	installSaveStorageTraits = f
 	return func() {
 		installSaveStorageTraits = old
 	}
 }
 
+func MockMatchDisksToGadgetVolumes(f func(gVols map[string]*gadget.Volume,
+	volCompatOpts *gadget.VolumeCompatibilityOptions) (map[string]map[int]*gadget.OnDiskStructure, error)) (restore func()) {
```
Comment (Meulengracht) [inline overlord/devicestate/export_test.go:415]: nitpick: `testutil.backup`?

File: gadget/ondisk.go (hunk 25 above)
Comment (alfonsosanchezbeato) [reply gadget/ondisk.go]: Changed

File: overlord/devicestate/export_test.go (hunk 26 above)
Comment (alfonsosanchezbeato) [reply overlord/devicestate/export_test.go:415]: Changed

**Review by alfonsosanchezbeato — COMMENTED** [review]
@Meulengracht thanks a lot for the review! I've addressed your comments now.

File: gadget/install/partition.go (hunk 23 above)
Comment (Meulengracht) [reply gadget/install/partition.go:171]: sorry I misread the code, my mistake! Sorry for the confusion

File: gadget/install/encrypt.go [hunk 27]
```diff
@@ -54,24 +53,24 @@ var _ = encryptedDevice(&encryptedDeviceLUKS{})
 
 // newEncryptedDeviceLUKS creates an encrypted device in the existing
 // partition using the specified key with the LUKS backend.
-func newEncryptedDeviceLUKS(part *gadget.OnDiskStructure, encType secboot.EncryptionType, key keys.EncryptionKey, label, name string) (encryptedDevice, error) {
+func newEncryptedDeviceLUKS(devNode string, encType secboot.EncryptionType, key keys.EncryptionKey, label, name string) (encryptedDevice, error) {
+	encLabel := label + "-enc"
+	if err := secbootFormatEncryptedDevice(key, encType, encLabel, devNode); err != nil {
```
[the remark this answers is not in the document]
Comment (Meulengracht) [reply gadget/install/encrypt.go:58]: Ah, I was pondering if it was an order thing, because if it was I thought it might have warranted a comment

**Review by Meulengracht — APPROVED** [review]
LGTM
